TypeScript (#1819)
enkidevs/curriculum: typescript/introduction/intro-typescript/what-is-typescript.md

---
author: kapnobatai136

type: normal

category: must-know

aspects:
  - introduction

---

# Intro to TypeScript

---
## Content

Depending on how old you are, you might remember the time JavaScript was just a web development language. Due to its increased versatility, JavaScript is now used in many applications such as designing neural networks, power backends and many others.

> You're probably thinking 'Hey, I thought this was a TypeScript course!' yet here we are talking about JavaScript. Because of the high adoption rate, many frameworks (Angular) and libraries (React) were created that aid with development. 

Reviewer: These two sentences don't flow with one another.

Reviewer: Did you mean to move the "Because of the high..." into the end of the paragraph above this one?

Also, there's no need to have the split between framework/library when mentioning Angular or React here. The distinction isn't particularly useful. Especially for learning TypeScript.

 You can use some more neutral less political/pedantic terms like "tool" or "project".

> TypeScript is another creation meant to put the future of JavaScript in the hands of developers, and it actually is a superset of JavaScript. What it means is that TypeScript retains all the functionality and adds more.

Reviewer: This is a very important feature of TypeScript. In fact, the #1 reason for its popularity is probably this interoperability with JavaScript. Converting js to ts is done by just changing the file extension.

```suggestion
TypeScript is another creation meant to put the future of JavaScript in the hands of developers, and it actually is a superset of JavaScript. What this means is that TypeScript retains all the functionality and adds more.
```

---
## Practice

TypeScript is a ??? of JavaScript.

* superset
* subset
* extraset

---
## Revision

> TypeScript retains ??? the functionality of JavaScript and ???.
> 
> * all
> * adds more
> * none of

Reviewer: ```suggestion
Any valid JavaScript is also valid TypeScript.

* true
* false
* not always, it depends
```

---
author: kapnobatai136

type: normal

category: must-know

aspects:
  - introduction

---

# Intro to TypeScript

---
## Content

If you're old enough, you might remember when JavaScript was just a web development language. Due to its versatility, JavaScript is now used in many applications including neural network design, power backends and many more.

You're probably thinking "What's JavaScript got to do with it -- isn't this a TypeScript course?". Let us explain!

Because of the high adoption rate of JavaScript, many tools (such as Angular, React, Node.js, etc) were created to allow programmers to write better/safer JavaScript code. 

TypeScript is another creation meant to put the future of JavaScript in the hands of developers, and it's actually a **superset of JavaScript**. What that means is that TypeScript has all the functionality of JavaScript, but *adds even more*.

> 

Reviewer: We can add an image here?

[typescript-is-superset-of-javascript](https://img.enkipro.com/074c0352be386e58bbdd697ed8bff427.png)

```suggestion

[typescript-is-superset-of-javascript](https://img.enkipro.com/074c0352be386e58bbdd697ed8bff427.png)
```

---
## Practice

Any valid JavaScript is also valid TypeScript.

???

* True.
* False.
* Not always, it depends.

---
## Revision

> TypeScript is a ??? of JavaScript.
> 
> * superset
> * subset
> * extraset

Reviewer: Focusing on the words "superset" and "subset" isn't as important as on the meaning behind them.

```suggestion
All features of JavaScript are included in TypeScript but not all features of TypeScript are included in JavaScript.

* true
* false
```

---
author: kapnobatai136

type: normal

category: must-know

aspects:
  - introduction

---

# Intro to TypeScript

---
## Content

> If you're old enough, you might remember when JavaScript was just a web development language. Due to its versatility, JavaScript is now used in many applications including neural network design, power backends and many more.

Reviewer 2: TIL there's TensorFlow.js; jesus we've come far

You're probably thinking "What's JavaScript got to do with it -- isn't this a TypeScript course?". Let us explain!

> Because of the high adoption rate of JavaScript, many tools (such as Angular, React, Node.js, etc) were created to allow programmers to write better/safer JavaScript code. 

Reviewer 2: I wouldn't say `Angular, React, Node` allowed `better/safer` code to be written; It's rather they allowed more specialized code to be written, that adhere to a certain programming style (in case of React/Angular) or allow JavaScript to be run on any computer, not just in browsers

Reviewer 3: Better examples would be Dart, Flow, Kotlin JS (Which is BAE btw)

Reviewer: Something like this @mihaiberq @sourabhv ?

```suggestion
Because of the high adoption rate of JavaScript, many tools were created to allow programmers to write code on multiple platforms, specialized for various programming styles. 
```

Reviewer 3: What does *write code on multiple platforms*? And *specialized for various programming styles*? Maybe explain these in insight briefly?

Reviewer: How about this? cc @sourabhv 

```suggestion
Because of the popularity of JavaScript, many tools were created to further support different programming patterns or environments. Some examples are tools that aim to improve web app development [such as Angular](https://angular.io/), enable mobile app development (such as [React Native](https://reactnative.dev/), or enable building backend servers (such as [Node.js](https://nodejs.dev/)).
```

Reviewer 3: Looks good, but since this is TS, an extension/transpiler of JS not implementation, we should also give examples of transpilers, not frameworks built on top of JS

Reviewer 3: Thats why I mentioned Dart Kotlin JS and Flow, More would be Coffeescript, Babel,

TypeScript is another creation meant to put the future of JavaScript in the hands of developers, and it's actually a **superset of JavaScript**. What that means is that TypeScript has all the functionality of JavaScript, but *adds even more*.

![typescript-is-superset-of-javascript](https://img.enkipro.com/074c0352be386e58bbdd697ed8bff427.png)

---
## Practice

Any valid JavaScript is also valid TypeScript.

???

> * True.

Reviewer 3: https://www.typescriptlang.org/play/#code/DYUwLgBAhhC8EHIASBLBBuAUDeBGX6QA

> * True.

Reviewer: That's just a matter of compiler flags.

Syntactically, every JS file is a valid TS file, although compilation might produce errors for some, depending on the compiler config flags.

[Here's a good explanation](https://stackoverflow.com/a/44967932)

> * True.

Reviewer: That being said, I just this could still lead to confusion.

Maybe we should say:

> Any valid JavaScript syntax is also valid TypeScript syntax

Although then we'd also need to introduce what "syntax" is.

> * True.

Reviewer 4: either way, I don't think we should be this vocal about this detail - I think there are many nuances to the statement which we can't properly describe unless we go into the details of the language:

a) in the snippet you shared they talk about programs and not valid programs; isn't this a big difference?
b) we have a seeming counterexample down below: https://github.com/enkidevs/curriculum/pull/1819/files/78e43d2ba6a665799a5199ef305b2f52aa85103a#diff-c4113eb1d2839c3f32285073c91d3fa1
c) we don't mention anything about compiler flags - if we just assume the default ones then the first impression is that js programs directly migrated to typescript can fail and throw errors -> that makes the program invalid?

how would we phrase this nuance? typescript can raise errors, but if we really want to run the compiled program it will work? or is it valid from JS pov, but not TS pov?
at which point is a program a typescript program - post compilation? doesn't make sense

> * True.

Reviewer 4: to make that statement true we have to say something like:
- well typescript errors can refer to a) errors that make the program invalid b) errors about the typing of the code, but don't make your program invalid 

if you program only has b) then your program can still be run by typescript

I can't stop thinking that this will just confuse users

> * True.

Reviewer: Agreed.
